### ISSUE TITLE
TFLite converter produces wrong output shape_signature if RNN / LSTM output layer in model (becomes all unknown dimensions)

### ISSUE THREAD

**Issue by shahruk10** [tensorflow/tensorflow#47207]
### 1. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- TensorFlow installation (pip package or built from source): pip
- TensorFlow library (version, if pip package or github SHA, if built from source): tf-nightly==2.5.0-dev20210216 and tensorflow>=2.4.0

### 2. Code

- [Collab notebook](https://colab.research.google.com/drive/1csQPhZpdidop2OS3JOGEGtCTU1JWwiWU?usp=sharing)

- Code:
```py
import tensorflow as tf
from pprint import pprint

def convertModelToTFLite(path, outPath):

    print(f"[INFO] Using tensorflow v{tf.__version__}")
    converter = tf.lite.TFLiteConverter.from_saved_model(path)
    converter.experimental_new_converter = True
    tfliteModel = converter.convert()

    # Save the model
    with open(outPath, 'wb') as f:
        f.write(tfliteModel)

units = 256
savePath = "my_model"

# creating model
i = tf.keras.Input(shape=(1, 521), name='input')
x = tf.keras.layers.Dense(units)(i)
x = tf.keras.layers.LSTM(units, return_sequences=True)(x)
# x = tf.keras.layers.Dense(units)(x)  # <-------- Having a dense output layer gives correct  output signature 

# saving as SavedModel
model = tf.keras.models.Model(inputs=[i], outputs=[x])
model.save(savePath, save_format='tf')

convertModelToTFLite(savePath, f"{savePath}.tflite", args.optimize)
ip = tf.lite.Interpreter(f"{savePath}.tflite")

print(model.summary())
pprint(ip.get_output_details())
```

### 3. Failure after conversion

- Conversion **is** successful and model **works** if I resize the output tensor to expected dimensions and reallocate tensors
- **But** I expect the shape signature of the output to be correct, and have the last dimension (feature dimension) be known and equal to the number of units / cells in the LSTM.

- With an LSTM or RNN layer as the output layer to a `keras.Model`, the TFLite model has the unexpected output shape signature :
```
  'shape_signature': array([-1, -1, -1], dtype=int32),      <---------- All -1s
``` 

- However with a Dense layer, the output has the expected shape signature:
```
  'shape_signature': array([ -1,  -1, 256], dtype=int32),
```

### 5. (optional) Any other info / logs
- Output details and model summary **with a RNN / LSTM output layer**

```
[{'dtype': <class 'numpy.float32'>,
  'index': 42,
  'name': 'StatefulPartitionedCall:0',
  'quantization': (0.0, 0),
  'quantization_parameters': {'quantized_dimension': 0,
                              'scales': array([], dtype=float32),
                              'zero_points': array([], dtype=int32)},
  'shape': array([1, 1, 1], dtype=int32),
  'shape_signature': array([-1, -1, -1], dtype=int32),      <---------- All -1s
  'sparsity_parameters': {}}]

___________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input (InputLayer)           [(None, 1, 521)]          0         
_________________________________________________________________
dense_3 (Dense)              (None, 1, 256)            133632    
_________________________________________________________________
lstm_3 (LSTM)                (None, 1, 256)            525312    
=================================================================
```

- Output details and model summary **with Dense output layer**

```
[{'dtype': <class 'numpy.float32'>,
  'index': 51,
  'name': 'StatefulPartitionedCall:0',
  'quantization': (0.0, 0),
  'quantization_parameters': {'quantized_dimension': 0,
                              'scales': array([], dtype=float32),
                              'zero_points': array([], dtype=int32)},
  'shape': array([  1,   1, 256], dtype=int32),
  'shape_signature': array([ -1,  -1, 256], dtype=int32),
  'sparsity_parameters': {}}]

_________________________________________________________________
Layer (type)                 Output Shape              Param #
=================================================================

input (InputLayer)           [(None, 1, 521)]          0
_________________________________________________________________
dense (Dense)                (None, 1, 256)            133632
_________________________________________________________________
lstm (LSTM)                  (None, 1, 256)            525312
_________________________________________________________________
dense_1 (Dense)              (None, 1, 256)            65792
=================================================================
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.4](https://colab.research.google.com/gist/amahendrakar/2d7fea304b1ddf2acef0a27816c9d40c/47207.ipynb) and TF-nightly.

Whereas with [TF v2.3](https://colab.research.google.com/gist/amahendrakar/a02ee07350673687a1e976aa70e8c285/47207.ipynb), I am facing an error while converting the model to `tflite` format. Please check the linked gist for reference. Thanks!

**Comment by teijeong** [comment]
Seems fixing the input size, as described in this [colab](https://colab.research.google.com/github/tensorflow/tensorflow/blob/master/tensorflow/lite/examples/experimental_new_converter/Keras_LSTM_fusion_Codelab.ipynb) helps.

```python
run_model = tf.function(lambda x: model(x))
# This is important, let's fix the input size.
BATCH_SIZE = 1
INPUT_SIZE = 521
concrete_func = run_model.get_concrete_function(
    tf.TensorSpec([BATCH_SIZE, 1, INPUT_SIZE], model.inputs[0].dtype))
model.save(savePath, save_format='tf', signatures=concrete_func)
```

**Comment by shahruk10** [comment]
> Seems fixing the input size, as described in this [colab](https://colab.research.google.com/github/tensorflow/tensorflow/blob/master/tensorflow/lite/examples/experimental_new_converter/Keras_LSTM_fusion_Codelab.ipynb) helps.
> 
> ```python
> run_model = tf.function(lambda x: model(x))
> # This is important, let's fix the input size.
> BATCH_SIZE = 1
> INPUT_SIZE = 521
> concrete_func = run_model.get_concrete_function(
>     tf.TensorSpec([BATCH_SIZE, 1, INPUT_SIZE], model.inputs[0].dtype))
> model.save(savePath, save_format='tf', signatures=concrete_func)
> ```

Yeah, if I define all dimensions of the input shape, it works. But if I leave either the batch size or timestep size as `None`, the problem returns. 

Also I realized, at least, after the first call to `interpreter.invoke()`, the output dimensions become expected values, but the signature dims remain (-1, -1, -1).

**Comment by shahruk10** [comment]
Fixed in [Tensorflow v2.5.0](https://github.com/tensorflow/tensorflow/releases/tag/v2.5.0), thanks !
